aten/src/ATen/native/Distributions.cpp

```diff
@@ -21,44 +21,67 @@ Tensor& bernoulli_(Tensor& self, double p, Generator* generator) {
 // TODO Replace this with more accurate digamma().
 template <typename scalar>
 static inline scalar digamma_one(scalar x) {
-  const scalar eps = x * 1e-2;
+  const scalar eps = x * 1e-3;
   return (std::lgamma(x + eps) - std::lgamma(x - eps)) / (eps + eps);
 }
 
-/** Computes the reparameterized gradient -(d/dalpha cdf(x;alpha)) / pdf(x;alpha)
-    for random number x drawn from a standard Gamma distribution Gamma(alpha).
-*/
+// Computes the reparameterized gradient -(d/dalpha cdf(x;alpha)) / pdf(x;alpha)
+// for random number x drawn from a standard Gamma distribution Gamma(alpha).
 template <typename scalar>
 static inline scalar standard_gamma_grad_one(scalar alpha, scalar x) {
-  // Use an asymptotic approximation for small x.
-  if (x < 0.2f) {
-    const auto a0 = 1 / alpha;
-    const auto a1 = 1 / (alpha + 1);
-    const auto a2 = 1 / (alpha + 2);
+  // Use a Taylor series expansion for small x.
+  if (x < 0.8f) {
+    scalar numer = 1;
+    scalar denom = alpha;
+    auto series1 = numer / denom;
+    auto series2 = numer / (denom * denom);
+    for (int i = 1; i <= 5; ++i) {
+      numer *= -x / i;
+      denom += 1;
+      series1 += numer / denom;
+      series2 += numer / (denom * denom);
+    }
     const auto pow_x_alpha = std::pow(x, alpha);
     const auto gamma_pdf = std::pow(x, alpha - 1) * std::exp(-x);
-    const auto gamma_cdf = pow_x_alpha * (a0 - x*a1 + 0.5f*x*x*a2);
+    const auto gamma_cdf = pow_x_alpha * series1;
     const auto gamma_cdf_alpha = (std::log(x) - digamma_one(alpha)) * gamma_cdf
-        - pow_x_alpha * (a0*a0 - x*a1*a1 + 0.5f*x*x*a2*a2);
+        - pow_x_alpha * series2;
     const auto result = -gamma_cdf_alpha / gamma_pdf;
     return std::isnan(result) ? 0 : result;
   }
 
-  // Use an asymptotic approximation for large alpha.
-  if (alpha > 50.0f) {
-    return std::sqrt(x / alpha);
+  // Use a Rice saddle point expansion for large alpha.
+  if (alpha > 8.0f) {
+    if (0.995f * alpha <= x && x <= 1.005f * alpha) {
+      const auto numer_1 = 1 + 24 * alpha * (1 + 12 * alpha);
+      const auto numer_2 = 1440 * (alpha * alpha) + 6 * x * (53 - 120 * x)
+          - 65 * x * x / alpha + alpha * (107 + 3600 * x);
+      const auto denom = 1244160 * (alpha * alpha) * (alpha * alpha);
+      const auto result = numer_1 * numer_2 / denom;
+      return std::isnan(result) ? 0 : result;
+    }
+    const auto denom = std::sqrt(8 * alpha);
+    const auto term2 = denom / (alpha - x);
+    const auto term3 = std::pow(x - alpha - alpha * std::log(x / alpha), -1.5);
+    const auto term23 = (x < alpha) ? term2 - term3 : term2 + term3;
+    const auto term1 = std::log(x / alpha) * term23
+                     - std::sqrt(2 / alpha) * (alpha + x) / ((alpha - x) * (alpha - x));
+    const auto stirling = 1 + 1 / (12 * alpha) * (1 + 1 / (24 * alpha));
+    const auto numer = x * term1;
+    const auto result = -stirling * numer / denom;
+    return std::isnan(result) ? 0 : result;
```

Comment:
why do we want to do this nan check?

Reply:
Hmm I guess I could omit it for large `alpha`. The small `x` check above is needed since the `Gamma` distribution puts lots of mass very close to 0 for small `alpha`, and I was just being paranoid here.

Reply:
Ok I've removed the two nan checks that haven't been a problem in practice; the third is indeed needed for very small x.